### ISSUE TITLE
Feature request - View's fields grouped by tab

### ISSUE THREAD

**Issue by jcodewalker** [vdm-io/Joomla-Component-Builder#411]
Hi Llewellyn,

I don't know if this is a need only mine... but what would you say to manage the view's field with nested subforms?

Example: in a view I set (let's say) 5 tabs and today I have to set all fields in just 1 subform (whichever is the tab they are referred to). There, I have to set db fields, backend config fields and even menu item(s) options, not to mention notes and spacer fields.

With complex project it's easy (for my works) to have something like 60 or more fields. And when the customer asks for a change it is a little bit more complex than what it could be.

What do you say to manage the view fields a little bit different?

In view setting, when adding tabs, we should be able to set if that tab is for backend (config/options), backend edit form, front-end form (which can be different from the backend) , menu item or the possible combinations between them (managed with checkboxes?).

Then, in view's fields tab, we should have a subform, let's call it tabs-container, with two fields: a list field containing all the tabs settled in view's setting tab we can choose from, and a subform field with the current options (less the tab the field is referred to, obviously).

In this way all the fields are grouped by tab, and each tab has its own display setting.
More over, I can use the same field for different scope and different project. I can have need to use a field in the backend only in a project but the need to use the same field in the front end in another view or project. Today I need to have the same field multiple times, with different extra-property-display setting. If something changes (with the project or Joomla's code) I have to manage the same field for all the "copies". In a brief time I have tons of identical fields with different display setting.

To refine more the field/tab/display options management, I would say to allow view's setting to manage tabs for view's backend edit form, front end form and menu item display only (aren't we in the view environment?).

Then, in project setting tab, we should have a new button to add tabs (for component's option/config display only), and a new 'config' tab (or 'options' tab) where to set the fields for component config display only.  

In this way I would say JCB would reflect even more what a Joomla developer is already used to, improving significantly the learning process.

What do you say?
Thanks.

**Comment by Llewellynvdm** [comment]
I actually do not follow your explanation. So I would ask that you give screenshots or a video tutorial explaining this.

Before you do, it could be that JCB already does what you are looking for and so you may need to learn the correct approach... Then even more to change how JCB links fields are not that simple as moving a few fields around in the GUI. There are many systems build upon the current implementation, systems that you may not even be aware exist at this time. Since JCB is a treasure trove, as you study the compiler you will be dumb struck at all the dynamic functions (basic AI). Remember the majority of the tutorials made up until now just cover the basic concepts of JCB :)

But okay please give some screen shots and clear drawings of the change you think will be helpful and then we can look at this again. The reason for this, is the fact that so many things in JCB and Joomla has the same names in different areas, which makes it hard to communicate in plain text documentation. That is why I make tutorials... this gets everyone on the same page much faster.

**Comment by jcodewalker** [comment]
I understand you at 100%, believe me. I was concerned to be perceived as not respecting all the work behind JCB but, believe it or not, I'm very aware of it. At the same time the more the project is complex the more JCB can help.

So here we are:
![tabs-management-explain](https://user-images.githubusercontent.com/49115855/56847447-541faf80-68db-11e9-9520-32d849105b55.jpg)
This picture is how I intend the fields tab to be.

If it makes sense for you too, I'll do the other pictures to full explain.

If JCB already does it... GREAT! lead me to the right video, please. 
Thanks.

**Comment by Llewellynvdm** [comment]
Okay you build the tabs on the adminn view and add the tabs to the fields... hmm that is how it works now.
![image](https://user-images.githubusercontent.com/5607939/56847603-73b7d780-68dd-11e9-9638-03f9f03007bb.png)

![image](https://user-images.githubusercontent.com/5607939/56847608-84684d80-68dd-11e9-8886-e2a073b3a2f9.png)

So you see the field called `admin tabs` is the selection option for the tabs.

Okay so please explain more :) if  you are going another direction with this. At this point the only benefit is less selection, but the down side is huge changes to the compiler, so I will need some more motivation here.

**Comment by jcodewalker** [comment]
I totally understand you and thanks for you attention.

Ok, let's see if I have completed the process.

admin view: create new tab:
![view-new-tab-create](https://user-images.githubusercontent.com/49115855/56848513-6523ed00-68ea-11e9-9dfd-e7629b91701c.jpg)
Here we decide where the tab will be displayed (not the single field).

Component setting tab:
![component-config-tab-create](https://user-images.githubusercontent.com/49115855/56848538-96042200-68ea-11e9-832c-f99ea07d74a1.jpg)
New button to create/add new config tab.

Component create new config tab process:
![component-config-new-tab-create](https://user-images.githubusercontent.com/49115855/56848548-b633e100-68ea-11e9-84f4-d8a9ef8a180f.jpg)
As this tab will only be displayed in component's config view doesn't require any more control.

Now, if I'm not wrong, fields do not require the display option anymore as they can be displayed anywhere depending on the tab they are included in.

Component's config are where a Joomla developer would look in, in component configuration environment. The order is determined by the tabs ordering.

View's fields are grouped by tab and there we can have a right idea of what is displayed where.
As the fields are now display independent, we can use the same field in more tabs which are displayed in different way.

Does it make sense?

**Comment by Llewellynvdm** [comment]
Sorry @jcodewalker I am afraid this is not going to happen anytime soon.

I appreciate your suggestions, but with relation to this one there is no real major improvement to JCB. All the features already exist, just in a different way. You should also realize that moving these at this stage will inevitably cause bugs and other pains towards those who have been using it in the way it already works.

So lets go over this, the config fields are set per component, and that may be a pain, but is really the best way as its implementation is able to reach down into the menu of a site view, and even any other area in your component. That being said, the idea to have config that is linked to a view and therefore added to any component that uses that view is not bad. So we keep that idea in the shelf and see if we can get that done after J4 and all that comes with those changes.

Then the option to tweak the show of fields with relation the the config, on the component level can be handy but hey doing this on a field level as it is now, is not that much worse. Reality is that you can very easy copy a field and change its attributes for a second implementation.

Then when working with custom fields, you can use one custom code placeholder between various versions of the same custom code. Secondly there is a prime switch in the custom fields that allows you to set one of those fields as the prime, to insure its code is what gets used for all fields with the same custom field type.

Then their is the whole reality of a custom tab implementation, which already has huge powerful features.

Then we need to really think about individual tab permissions, as we already have permissions down the the field level, yet the idea to add a permission switch to tabs is actually the one idea that I think we can explore. But it also will need to take a back seat... as there are some other more important stuff in the todo list for now. Yet if I get the time... I will bump that in :-1:

So in short config will for now stay in the component view as it is, and the switch to turn off the field from showing in the global options or just the menu options is done on field level. Let me know if you don't know how this is done. Just that you know if you use that field in a admin view as well as in the config, they are independent and the show in global options or menu is just in relation to the config fields, and ignored when the same field is used in a normal admin view.

I am closing this issue, but you are welcome to continue the communication about this over here, and would I add any of these ideas I will reference this issue :+1:

**Comment by jcodewalker** [comment]
Thank you for your time, first of all.

I understand everything and even all your concerns, so keep it up, bro.

Have anice sunday.
claudio
